tests/deployments/salt/salt_test.py

```diff
@@ -0,0 +1,119 @@
+
+CONFIG = """
+signalfx-agent:
+  package_stage: null
+  version: null
+  conf:
+    signalFxAccessToken: 'testing123'
+    ingestUrl: null
+    apiUrl: null
+    intervalSeconds: 1
+    observers:
+      - type: host
+    monitors: null
+"""
+
+PILLAR_PATH = "/srv/pillar/signalfx-agent.sls"
+SALT_CMD = "salt-call --local state.apply"
+STAGE = os.environ.get("STAGE", "final")
+INITIAL_VERSION = os.environ.get("INITIAL_VERSION", "4.7.5-1")
+UPGRADE_VERSION = os.environ.get("UPGRADE_VERSION", "4.7.6-1")
+
+
+def get_config(backend, agent_version, monitors, stage):
+    config_yaml = yaml.safe_load(CONFIG)
+    config_yaml["signalfx-agent"]["package_stage"] = stage
+    config_yaml["signalfx-agent"]["version"] = agent_version
+    config_yaml["signalfx-agent"]["conf"]["ingestUrl"] = backend.ingest_url
+    config_yaml["signalfx-agent"]["conf"]["apiUrl"] = backend.api_url
+    config_yaml["signalfx-agent"]["conf"]["monitors"] = monitors
+    return yaml.dump(config_yaml)
+
+
+def run_salt(cont, backend, agent_version, monitors, stage):
+    with tempfile.NamedTemporaryFile(mode="w+") as fd:
+        config_yaml = get_config(backend, agent_version, monitors, stage)
+        print(config_yaml)
+        fd.write(config_yaml)
+        fd.flush()
+        copy_file_into_container(fd.name, cont, PILLAR_PATH)
+    code, output = cont.exec_run(SALT_CMD)
+    assert code == 0, output.decode("utf-8")
+    print_lines(output)
+    agent_version = agent_version.rstrip("-1").rstrip("~1")
+    installed_version = get_agent_version(cont).rstrip("-1").rstrip("~1")
+    assert installed_version == agent_version, "installed agent version is '%s', expected '%s'" % (
+        installed_version,
+        agent_version,
+    )
+    assert is_agent_running_as_non_root(cont), "Agent is not running as non-root user"
+
+
+@pytest.mark.parametrize(
+    "base_image,init_system",
+    [pytest.param(distro, init, marks=pytest.mark.deb) for distro, init in DEB_DISTROS]
+    + [pytest.param(distro, init, marks=pytest.mark.rpm) for distro, init in RPM_DISTROS],
+)
+def test_salt(base_image, init_system):
+    opts = {"path": REPO_ROOT_DIR, "dockerfile": DOCKERFILES_DIR / f"Dockerfile.{base_image}", "with_socat": False}
+    with run_init_system_image(base_image, **opts) as [cont, backend]:
+        try:
+            monitors = [{"type": "host-metadata"}]
+            run_salt(cont, backend, INITIAL_VERSION, monitors, STAGE)
+            assert wait_for(
+                p(has_datapoint_with_dim, backend, "plugin", "host-metadata")
+            ), "Datapoints didn't come through"
+
+            # upgrade agent
+            run_salt(cont, backend, UPGRADE_VERSION, monitors, STAGE)
+            assert wait_for(
+                p(has_datapoint_with_dim, backend, "plugin", "host-metadata")
+            ), "Datapoints didn't come through"
+
+            # downgrade agent
+            run_salt(cont, backend, INITIAL_VERSION, monitors, STAGE)
+            assert wait_for(
```

Comment:
Definitely want to clear the datapoints in the backend before waiting for new ones, otherwise this test could be easily fooled.